### PR TITLE
etcdserver/api/v3rpc: fill in KV ResponseHeader

### PR DESCRIPTION
**Pull request by gyuho**

### CODE REVIEW

File: etcdserver/api/v3rpc/key.go [hunk 1]
```diff
@@ -55,6 +64,7 @@ func (s *kvServer) Put(ctx context.Context, r *pb.PutRequest) (*pb.PutResponse,
 		return nil, togRPCError(err)
 	}
 
+	resp.Header = s.newResponseHeader(resp.Header.Revision)
```
Comment (xiang90) [inline etcdserver/api/v3rpc/key.go]: i do not quite understand this line. Why do we assign resp.Header.Revision back to resp.Header.Revision?

Comment (gyuho) [inline etcdserver/api/v3rpc/key.go]: @xiang90 `s.kv.Put` already populates `Revision` in the header with other member vars zero.
So I just wanted to populate header from `kvServer` and overwrite `resp.Header`. It's ugly, I can
change it. Any suggestion?

**Comment by gyuho** [comment]
@xiang90 Do you think this approach is correct? Please let me know if you have feedback on the logic overall. Thanks,

File: etcdserver/api/v3rpc/key.go (hunk 1 above)
Comment (xiang90) [inline etcdserver/api/v3rpc/key.go]: Then why not just do s.NewResponseHeader without that additional arg?

**Comment by gyuho** [comment]
@xiang90 Changed it to fill in values without overwriting function. PTAL. Thanks!

File: etcdserver/api/v3rpc/key.go [hunk 2]
```diff
@@ -42,6 +50,11 @@ func (s *kvServer) Range(ctx context.Context, r *pb.RangeRequest) (*pb.RangeResp
 		return nil, togRPCError(err)
 	}
 
+	if resp.Header != nil {
```
Comment (xiang90) [inline etcdserver/api/v3rpc/key.go]: we should probably panic here if err is not nil but header is nil

Comment (gyuho) [inline etcdserver/api/v3rpc/key.go]: Sure

File: etcdserver/api/v3rpc/key.go [hunk 3]
```diff
@@ -42,6 +50,11 @@ func (s *kvServer) Range(ctx context.Context, r *pb.RangeRequest) (*pb.RangeResp
...
+		resp.Header.ClusterId = uint64(s.clusterID)
```
Comment (xiang90) [inline etcdserver/api/v3rpc/key.go]: how about adding a func like s.FillInHeader(h *header)?

Comment (gyuho) [inline etcdserver/api/v3rpc/key.go]: Ok this sounds better, just passing pointer. Will change. Thanks,

**Comment by gyuho** [comment]
@xiang90 Please take a look again. Just added `fillInHeader` method and added panics when `resp.Header` is `nil` when `err` is `nil`.

File: etcdserver/api/v3rpc/key.go [hunk 4]
```diff
@@ -42,6 +55,10 @@ func (s *kvServer) Range(ctx context.Context, r *pb.RangeRequest) (*pb.RangeResp
 		return nil, togRPCError(err)
 	}
 
+	if resp.Header == nil {
+		plog.Panicf("got nil resp.Header")
```
Comment (xiang90) [inline etcdserver/api/v3rpc/key.go]: unexpected nil resp.Header

File: etcdserver/api/v3rpc/key.go [hunk 5]
```diff
@@ -90,9 +119,21 @@ func (s *kvServer) Compact(ctx context.Context, r *pb.CompactionRequest) (*pb.Co
...
+	}
+	s.fillInHeader(resp.Header)
 	return resp, nil
 }
 
+// fillInHeader populates pb.ResponseHeader's members
```
Comment (xiang90) [inline etcdserver/api/v3rpc/key.go]: please make sure you reach 80-100 chars before create a new line.

Comment (gyuho) [inline etcdserver/api/v3rpc/key.go]: Ok will change.

**Comment by gyuho** [comment]
@xiang90  I just fixed those panic message and function description. PTAL. Thanks,

File: etcdserver/api/v3rpc/key.go [hunk 6]
```diff
@@ -24,12 +25,24 @@ import (
...
+var (
+	plog = capnslog.NewPackageLogger("github.com/coreos/etcd/etcdserver/api", "v3rpc")
+)
+
 type kvServer struct {
-	kv etcdserver.RaftKV
+	clusterID int64
+	memberID  int64
+	raftTimer etcdserver.RaftTimer
+	kv        etcdserver.RaftKV
```
Comment (xiang90) [inline etcdserver/api/v3rpc/key.go]: a blank line between metadata and kv?

**Comment by xiang90** [comment]
LGTM

File: etcdserver/api/v3rpc/key.go (hunk 6 above)
Comment (gyuho) [inline etcdserver/api/v3rpc/key.go]: Ok will add a blank line. Thanks

**Comment by gyuho** [comment]
Just added blank line to divide as:

```
type kvServer struct {
	clusterID int64
	memberID  int64
	raftTimer etcdserver.RaftTimer

	kv etcdserver.RaftKV
}
```

Will merge after CI passes. Thanks,

**Pull request merged**